controller/authorize_blackbox_test.go

```diff
@@ -0,0 +1,301 @@
+	authorizeCtx, err = app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.NotNil(t, err)
+	require.Contains(t, err.Error(), "400 invalid_request: missing required parameter")
+	jsonapi.JSONErrorResponse(authorizeCtx, err)
+	require.Equal(t, 400, rw.Code)
+}
+
+func (rest *TestAuthorizeREST) TestAuthorizeUnauthorizedError() {
+	t := rest.T()
+	svc, ctrl := rest.UnSecuredController()
+
+	redirect := "https://openshift.io"
+	clientID := ""
+	responseType := "code"
+	state := uuid.NewV4()
+
+	test.AuthorizeAuthorizeUnauthorized(t, svc.Context, svc, ctrl, nil, clientID, redirect, responseType, nil, state)
+}
+
+func (rest *TestAuthorizeREST) TestAuthorizeCallbackOK() {
+	t := rest.T()
+	_, ctrl := rest.UnSecuredController()
+
+	rw := httptest.NewRecorder()
+	u := &url.URL{
+		Path: fmt.Sprintf(client.AuthorizeAuthorizePath()),
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	require.Nil(t, err)
+
+	redirectURI := "https://openshift.io/somepath"
+	prms := url.Values{}
+
+	prms.Add("response_type", "code")
+	prms.Add("redirect_uri", redirectURI)
+	prms.Add("client_id", rest.Configuration.GetPublicOauthClientID())
+	prms.Add("state", uuid.NewV4().String())
+
+	ctx := context.Background()
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	authorizeCtx, err := app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.Nil(t, err)
+
+	err = ctrl.Authorize(authorizeCtx)
+	require.Nil(t, err)
+
+	require.Equal(t, 307, rw.Code) // redirect to keycloak login page.
+
+	locationString := rw.HeaderMap["Location"][0]
+	locationUrl, err := url.Parse(locationString)
+	require.Nil(t, err)
+
+	allQueryParameters := locationUrl.Query()
+
+	require.NotNil(t, allQueryParameters)
+	require.NotNil(t, allQueryParameters["state"][0])
+
+	returnedState := allQueryParameters["state"][0]
+
+	u = &url.URL{
+		Path: fmt.Sprintf(client.CallbackAuthorizePath()),
+	}
+
+	prms = url.Values{
+		"state": {returnedState},
+		"code":  {"SOME_OAUTH2.0_CODE"},
+	}
+
+	ctx = context.Background()
+	rw = httptest.NewRecorder()
+
+	req, err = http.NewRequest("GET", u.String(), nil)
+	require.Nil(t, err)
+
+	// The OAuth code is sent as a query parameter by calling /api/login?code=_SOME_CODE_&state=_SOME_STATE_
+	// The request originates from Keycloak after a valid authorization by the end user.
+	refererKeycloakUrl := "https://keycloak-url.example.org/path-of-login"
+	req.Header.Add("referer", refererKeycloakUrl)
+
+	goaCtx = goa.NewContext(goa.WithAction(ctx, "AuthorizecallbackTest"), rw, req, prms)
+	callbackCtx, err := app.NewCallbackAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.Nil(t, err)
+
+	err = ctrl.Callback(callbackCtx)
+	require.Nil(t, err)
+
+	require.Equal(t, 307, rw.Code) // redirect to keycloak login page.
+
+	require.Contains(t, rw.Header().Get("Location"), redirectURI)
+
+	locationString = rw.HeaderMap["Location"][0]
```

Comment:
Verify that the locationString == the expected URL.

Reply:
Isn't ```require.Contains(t, rw.Header().Get("Location"), redirectURI)``` enough here?

Reply:
No, it's not enough. The location should be something like this: https://sso.openshift.io/auth/realm/....?redirect={redirectURI}
and you check if redirectURI is part of this URL. But we should also check if Location starts with https://sso.openshift.io/auth/realm/ or whatever is used in our test env. The URL to our Identity Provider which is Keycloak in this case.

Reply:
The location is redirectURI?code={code}&state={state} (which is correct as per the way I have implemented). If it should be the way you are suggesting I will have to change the implementation. Callback should redirect to redirectURI along with code and state. I don't see a reason to redirect to https://sso.openshift.io/auth/realm/....?redirect={redirectURI} .

Reply:
Ah, I didn't notice you asked about callback check. Yes, for callback URL it should be the original redirect. Not sso. SSO URL should be checked for authorize calls.

Reply:
But it would be better if you check that the URL string starts with the expected string. Not just contains it (because in this case it will pass even if it's just in a param).